src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
+            last.nextElement = current;
+        }
+        last = current;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        indexCheckerForAdding(index);
+        if (size() == DEFAULT_INDEX) {
+            add(value);
+            return;
+        }
+        if (index == DEFAULT_INDEX) {
+            Node<T> current = first;
+            first = new Node<>(value);
+            first.nextElement = current;
+            current.prevElement = first;
+            size++;
+            return;
+        }
+        if (index == size()) {
+            add(value);
+            return;
+        }
+        Node<T> current = getNodeByIndex(index);
+        Node<T> newNode = new Node<>(value);
+        newNode.prevElement = current.prevElement;
+        newNode.nextElement = current;
+        current.prevElement.nextElement = newNode;
+        current.prevElement = newNode;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T element : list) {
+            add(element);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        indexChecker(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexChecker(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecker(index);
+        Node<T> current = getNodeByIndex(index);
+        unlink(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if ((current.value == null && object == null)
+                    || (current.value != null && (current.value.equals(object)))) {
+                unlink(current);
+                return true;
+            }
+            current = current.nextElement;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == DEFAULT_INDEX;
+    }
+
+    private void indexChecker(int index) {
+        if (index < DEFAULT_INDEX || index >= size()) {
+            throw new IndexOutOfBoundsException("Index is out of bounds for size "
+                    + size() + " with element " + index);
+        }
+    }
+
+    private void indexCheckerForAdding(int index) {
+        if (index < DEFAULT_INDEX || index > size()) {
+            throw new IndexOutOfBoundsException("Index is out of bounds for size "
+                    + size() + " with element " + index);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index <= size() / DIVIDER) {
```

Comment:
we also can use `2` instead of `DIVIDER` 
it is clear that we check if index in the first or second half of our list